pkg/cloudprovider/provider/opennebula/provider.go

```diff
@@ -0,0 +1,430 @@
+}
+
+func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clusterv1alpha1.MachineSpec) error {
+	_, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %w", err)
+	}
+
+	opennebulaCloudProviderSpec := CloudProviderSpec{}
+	if err = json.Unmarshal(pc.CloudProviderSpec.Raw, &opennebulaCloudProviderSpec); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *provider) GetCloudConfig(_ clusterv1alpha1.MachineSpec) (string, string, error) {
+	return "", "", nil
+}
+
+func (p *provider) Create(ctx context.Context, _ *zap.SugaredLogger, machine *clusterv1alpha1.Machine, data *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c)
+
+	// build a template
+	tpl := vm.NewTemplate()
+
+	// add extra template vars first
+	for key, value := range c.VMTemplateExtra {
+		tpl.Add(keys.Template(key), value)
+	}
+
+	tpl.Add(keys.Name, machine.Spec.Name)
+	tpl.CPU(*c.Cpu).Memory(*c.Memory).VCPU(*c.Vcpu)
+
+	disk := tpl.AddDisk()
+	disk.Add(shared.Image, c.Image)
+	disk.Add(shared.Datastore, c.Datastore)
+	disk.Add(shared.Size, *c.DiskSize)
+
+	nic := tpl.AddNIC()
+	nic.Add(shared.Network, c.Network)
+	nic.Add(shared.Model, "virtio")
+
+	if c.EnableVNC {
+		tpl.AddIOGraphic(keys.GraphicType, "VNC")
+		tpl.AddIOGraphic(keys.Listen, "0.0.0.0")
+	}
+
+	tpl.AddCtx(keys.NetworkCtx, "YES")
+	tpl.AddCtx(keys.SSHPubKey, "$USER[SSH_PUBLIC_KEY]")
+
+	tpl.AddCtx(machineUidContextKey, string(machine.UID))
+	tpl.AddCtx("USER_DATA", base64.StdEncoding.EncodeToString([]byte(userdata)))
+	tpl.AddCtx("USER_DATA_ENCODING", "base64")
+
+	controller := goca.NewController(client)
+
+	// create VM from the generated template above
+	vmID, err := controller.VMs().Create(tpl.String(), false)
+	if err != nil {
+		return nil, err
+	}
+
+	vm, err := controller.VM(vmID).Info(false)
+	if err != nil {
+		return nil, err
+	}
+
+	return &openNebulaInstance{vm}, nil
+}
+
+func (p *provider) Cleanup(ctx context.Context, log *zap.SugaredLogger, machine *clusterv1alpha1.Machine, data *cloudprovidertypes.ProviderData) (bool, error) {
+	instance, err := p.Get(ctx, log, machine, data)
+	if err != nil {
+		if errors.Is(err, cloudprovidererrors.ErrInstanceNotFound) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return false, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c)
+	controller := goca.NewController(client)
+
+	vmctrl := controller.VM(instance.(*openNebulaInstance).vm.ID)
```

Comment:
I have to say I'm not a huge fan of the type assertions here and later in the code since it can introduce run-time panics. I understand it's required to get the OpenNebula specific information, but e.g. the AWS provider implements a private version of `get` to return the provider-specific instance: https://github.com/kubermatic/machine-controller/blob/df72929d4011f49d39f4f5f78208856ebba2ae99/pkg/cloudprovider/provider/aws/provider.go#L877

Have you considered that approach?

Reply:
No, this hasn't come across my mind, will try to adapt this then.